server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2001,7 +2001,7 @@ public void resetRecoveryStage() {
         if (state != IndexShardState.RECOVERING) {
             throw new IndexShardNotRecoveringException(shardId, state);
         }
-        recoveryState().setStage(RecoveryState.Stage.INIT);
+        this.recoveryState = recoveryState().reset();
```

Comment:
I wonder if we need to take the mutex around this? Did not find a specific issue though but seems we protect changes to the variable by `mutex`?

Reply:
I had the impression that we guard the `state` field with that mutex. But just to be safe I'll add the mutex guard here too.